modules/ssh/ssh.go

```diff
@@ -142,10 +143,12 @@ func sessionHandler(session ssh.Session) {
 	// Wait for the command to exit and log any errors we get
 	err = cmd.Wait()
 	if err != nil {
-		log.Error("SSH: Wait: %v", err)
+		if _, ok := err.(*exec.ExitError); !ok {
```

Comment:
What does this `err.(*exec.ExitError)` do? Check if the command has non-zero exit code?

Reply:
If `gitea serv` exits with a non-zero exit code - an `*exec.ExitError` will be returned by `cmd.Wait()`.

The contents of this error will be simply the exit code without any further context.

That is because the actual error that occurred will have been emitted on the stderr pipe and sent to the downstream user.

Now... 

There are many ways in which `gitea serv` could exit with a non-zero exit code as part of normal behaviour. Thus logging this error is pointless - especially as we can get no context for it.

However, `cmd.Wait()` can return with other errors, and we should log those. Therefore check the error and if it's not an `*exec.ExitError` log it - otherwise drop it.

Reply:
Ok. Would `errors.Is` work in this case, or is that an oddity of the `exec` API that one has to type assert like that?

Reply:
No. It would not work. `errors.Is` does an equality comparison on the target *NOT* a type comparison. If the target error implements `interface{ Is(error) bool }` and that did the correct type comparison we could use it but `exec.ExitError` does not do this.

<details><summary> `errors.Is(err, target error) bool`</summary>

```go
func Is(err, target error) bool {
	if target == nil {
		return err == target
	}

	isComparable := reflectlite.TypeOf(target).Comparable()
	for {
		if isComparable && err == target {
			return true
		}
		if x, ok := err.(interface{ Is(error) bool }); ok && x.Is(target) {
			return true
		}
		// TODO: consider supporting target.Is(err). This would allow
		// user-definable predicates, but also may allow for coping with sloppy
		// APIs, thereby making it easier to get away with them.
		if err = Unwrap(err); err == nil {
			return false
		}
	}
}
```

</details>


<details><summary> `type ExitError struct { ... }` </summary>

```go
// An ExitError reports an unsuccessful exit by a command.
type ExitError struct {
	*os.ProcessState

	// Stderr holds a subset of the standard error output from the
	// Cmd.Output method if standard error was not otherwise being
	// collected.
	//
	// If the error output is long, Stderr may contain only a prefix
	// and suffix of the output, with the middle replaced with
	// text about the number of omitted bytes.
	//
	// Stderr is provided for debugging, for inclusion in error messages.
	// Users with other needs should redirect Cmd.Stderr as needed.
	Stderr []byte
}

func (e *ExitError) Error() string {
	return e.ProcessState.String()
}
```

and `*os.ProcessState` does not have an `Is(error) bool` function either.

</details>

Reply:
It's a failure of the std library and the `errors.Is` function. All of these errors should be implementing `Is` and advertising in a clear and reasonable way what it means for an error to be `Is(...)` another error.

A present `errors.Is` is a massive gotcha.

Reply:
Yeah, error handling in golang stdlib is a huge mess, the error classes are not even named consistently, and what I'm really missing from other languages are exceptions, but I guess golang will never get them as they are already too invested in the error return values at this point.

Reply:
If this question already popped up during the review, I guess we can expect it to pop up more often.
What about a comment?

Reply:
```suggestion
		// Cannot use errors.Is here because ExitError doesn't implement Is
		// Thus errors.Is will do equality test NOT type comparison 
		if _, ok := err.(*exec.ExitError); !ok {
```

The problem is if we start adding comments like this we're going to need to add them everywhere we do error type checks.